tests/virtualization/universal/waitfor_guests.pm

```diff
@@ -102,15 +102,12 @@ sub run {
     add_guest_to_hosts $_, $virt_autotest::common::guests{$_}->{ip} foreach (keys %virt_autotest::common::guests);
     assert_script_run "cat /etc/hosts";
 
-    # Wait for guests to finish installation
-    # script_run("wait", timeout => 1200);
-    # script_retry("! ps x | grep -v 'grep' | grep 'virt-install'", retry => 120, delay => 10);
-    # Unfortunately this doesn't cover the second step of the installation, where ssh is available
-    my $sleep_delay = 1200;
-    $sleep_delay = 2400 if (is_xen_host);    # XEN has more guests
-    sleep($sleep_delay);    # XXX Get rid of this sleep!
-    wait_guest_online($_, 120) foreach (keys %virt_autotest::common::guests);
-    record_info("guests installed", "Guest installation completed");
+    # Wait for guests to announce that installation is complete
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        script_retry("cat guests_log | grep $guest", retry => 40, delay => 60);
+        record_info("$guest installed", "Guest installation completed");
+    }
```

Comment:
In this foreach code block, for each guest, it will maximumly use 40 minutes to declare failure. In fact only the first checked guest needs long timeout, from the second guest, since they are checked AFTER the first guest installation finishes, smaller timeout value is enough?

Reply:
Yes, I have this concern too. Given a extreme case, that all 10 guests failed, then we have to wait for 6 hours?

Reply:
@Julie-CAO 
If the first guest fail the test will fail in 40 minutes, if some of the other guests fail we will wait additional 40 minutes to catch that. So the total time to wait in case of failure is around 80 minutes.

@alice-suse 
Yes, this can create a small improvement in case of failure. We can wait for first successful guest install and run the foreach loop after that. In this case we have to  wait only the possible delta time between first guest install and the slowest one.

```
    script_retry("cat guests_log", retry => 40, delay => 60);
    foreach my $guest (keys %virt_autotest::common::guests) {
        script_retry("cat guests_log | grep $guest", retry => 10, delay => 60);
        record_info("$guest installed", "Guest installation completed");
    }
```

guests_log is created by the first installed guest, after that the remaining guest will have ~10 minutes to finish.

If we agree this is a better solution I can run some test and make a commit.

Reply:
Yes, that works. Or you can set the retry number differently for the first guest and the latter guests. Code change can be smaller?

Reply:
>If the first guest fail the test will fail in 40 minutes, if some of the other guests fail we will wait additional 40 minutes to catch that. So the total time to wait in case of failure is around 80 minutes.

oh, I did not notice that the test will fail in the case script_retry fails. Your current logic is to start multiple threads to install guests, once one of them is found to fail to output to guests_log, the test will abort at once? if so, only those completed ones which listed in front of the failing one are recorded by the test, other guests are in an unknow status? 

Given we install 5 guests, guest 1 ~ guest 5,  they are being installed in 5 virt-install sessions seperately.

firstly we check guest 1 in guests_log, it is good and flagged 'completed';

then we check guest 2, it is bad, the test aborts... we don't know the status of guest 3, 4 and 5.

Since all the guests are already been installed, guest 3, and 5 are most likely rebooted successfully, why do not we mark them 'completed' as well?

Reply:
> * Are there verification runs for failing guests?

Yes, I have made a run with simulated guest failure. I will put an example when I made changes to this PR.

> * Do the host hypervisor host and all guests have the identical pair of private key and public key? What are the benefits of doing this?

No specific reason, this is a key that is already public on github, so I decided to just use the same one to keep things simple.


When doing [poo#102984](https://progress.opensuse.org/issues/102984) the failed guests will have a chance to retry the installation. Here I will have to make some changes that can include information as "guest 1,4,5 are successful and guest 2 and 3 had fail".

Reply:
script_retry makes the check "either" pass or "die" so  that the guest checks must be done serially. We will change to check in parallel in the future so that we can mark  'completed' for the guests which finish installation/reboot successfully.  see https://progress.opensuse.org/issues/112559

Reply:
Julie will open a new ticket to track this suggestion. We will plan it in future.